File: 0115-distinct-subsequences/0115-distinct-subsequences.js
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var numDistinct = function(s, t) {
     const m = s.length;
    const n = t.length;

    const dp = new Array(m + 1).fill().map(() => new Array(n + 1).fill(0));

    for (let i = 0; i <= m; i++) {
        dp[i][n] = 1;
    }

    for (let sIndex = m - 1; sIndex >= 0; sIndex--) {
        for (let tIndex = n - 1; tIndex >= 0; tIndex--) {
            dp[sIndex][tIndex] = dp[sIndex + 1][tIndex];

            if (s[sIndex] === t[tIndex]) {
                dp[sIndex][tIndex] += dp[sIndex + 1][tIndex + 1];
            }
        }
    }

    return dp[0][0];
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.rb
# @param {String} s
# @param {String} t
# @return {Integer}
def num_distinct(s, t)
    m = s.length
    n = t.length

    dp = Array.new(m + 1) { Array.new(n + 1, 0) }

    (0..m).each do |i|
        dp[i][n] = 1
    end

    (m - 1).downto(0) do |s_index|
        (n - 1).downto(0) do |t_index|
            dp[s_index][t_index] = dp[s_index + 1][t_index]

            if s[s_index] == t[t_index]
                dp[s_index][t_index] += dp[s_index + 1][t_index + 1]
            end
        end
    end

    dp[0][0]
end